File: programmers/입국심사.py
"""
코딩테스트 연습 > 이분탐색 > 입국심사
"""


times = None


def get_count(time):
    result = 0
    for t in times:
        result += time // t
    return result


def solution(n, _times):
    global times
    times = _times

    left = 0
    right = max(times) * n
    while left <= right:
        mid = (left + right) // 2
        count = get_count(mid)
        if count < n:
            left = mid+1
        elif count > n:
            right = mid-1
        else:
            right = mid-1

    return left


print(solution(11, [4, 5]))

File: programmers/입국심사.kt
class Solution {
    fun solution(n: Int, times: IntArray): Long {
        fun evaluate(time: Long): Long {
            return times.map{time/it}.sum()
        }

        var start = 1L
        var end = 1000000000000000000L
        var mid = 0L
        while (start < end) {
            mid = (start+end)/2

            println("$start, $end, $mid, ${evaluate(mid)}")

            if (evaluate(mid) >= n) {
                end = mid
            } else {
                start = mid+1
            }
        }
        return end
    }
}
